cirq-google/cirq_google/named_topologies.py

```diff
@@ -0,0 +1,329 @@
+        to NetworkX plotting functionality.
+    """
+    if ax is None:
+        ax = plt.gca()  # coverage: ignore
+
+    if cartesian:
+        pos = {node: (y, -x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+    else:
+        pos = {node: (x + y, y - x) for node, (x, y) in _node_and_coordinates(graph.nodes)}
+
+    nx.draw_networkx(graph, pos=pos, ax=ax, **kwargs)
+    ax.axis('equal')
+    return pos
+
+
+@dataclass(frozen=True)
+class LineTopology(NamedTopology):
+    """A 1D linear topology.
+
+    Node indices are contiguous integers starting from 0 with edges between
+    adjacent integers.
+
+    Args:
+        n_nodes: The number of nodes in a line.
+    """
+
+    n_nodes: int
+
+    def __post_init__(self):
+        if self.n_nodes <= 1:
+            raise ValueError("`n_nodes` must be greater than 1.")
+
+    # https://github.com/python/mypy/issues/1362
+    @property  # type: ignore
+    @cache
+    def name(self) -> str:
+        """The name of this topology: {n_nodes}-line"""
+        return f'line-{self.n_nodes}'
+
+    @property  # type: ignore
+    @cache
+    def graph(self) -> nx.Graph:
+        """The graph of this topology.
+
+        Node indices are contiguous integers starting from 0 with edges between
+        adjacent integers.
+        """
+        return nx.from_edgelist(
+            [(i1, i2) for i1, i2 in zip(range(self.n_nodes), range(1, self.n_nodes))]
+        )
+
+    def draw(self, ax=None, cartesian: bool = True, **kwargs) -> Dict[Any, Tuple[int, int]]:
+        """Draw this graph.
+
+        Args:
+            ax: Optional matplotlib axis to use for drawing.
+            cartesian: If True, draw as a horizontal line. Otherwise, draw on a diagonal.
+            kwargs: Additional arguments to pass to `nx.draw_networkx`.
+        """
+        g2 = nx.relabel_nodes(self.graph, {n: (n, 1) for n in self.graph.nodes})
+        return draw_gridlike(g2, ax=ax, cartesian=cartesian, **kwargs)
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return dataclass_json_dict(self, namespace='cirq.google')
+
+
+@dataclass(frozen=True)
+class DiagonalRectangleTopology(NamedTopology):
+    """A grid topology forming a rectangle rotated 45-degrees.
+
+    This topology is based on Google devices where plaquettes consist of four qubits in a square
+    connected to a central qubit:
+
+        x   x
+          x
+        x   x
+
+    The corner nodes are not connected to each other. `width` and `height` refer to the number
+    of unit cells, or equivalently the number of central nodes. Each unit cell contributes
+    two nodes when in bulk: the central node and 1/4 of each of the four shared corner nodes.
+    Accounting for the boundary, the total number of nodes is 2*w*h + w + h + 1. An example
+    diagram showing the diagonal nature of the rectangle is reproduced below. It is a
+    "diagonal-rectangle-3-2" with width 3 and height 2. This can be most clearly seen by focusing
```

Comment:
I find this odd that this is central qubits only.  I would expect the below to be a rectangle of 4x3 not 3x2.  I'll defer this if others find this to be intuitive or if this is a standard in surface code / error correction, but I find it unnecessarily complicated and confusing to only include 'central qubits'

Reply:
I'm not sure there's a smaller [unit cell](https://en.wikipedia.org/wiki/Unit_cell) than this

Reply:
Are we sure this is the topology we want?  I find this definition very tricky to reason about and work with.  In particular, it's not actually a rectangle since a rectangle would have the same number of qubits in each diagonal.  It is definitely not what I thought it was initially when I first casually scanned it. 

Am I correct in saying that neither the Sycamore 54 or Bristlecone chip layout would satisfy being a DiagnonalRectangleTopology?  

I am willing to drop this if other maintainers feel that this is an intuitive definition, but I find it confusing and not what I expected.

Reply:
as far as I can tell, this is the shape of our chips. Willing to change the name if there's something more common, but we need this shape.

Reply:
I think SYCAMORE23 has this shape.  SYCAMORE 54 and Bristlecone do not though.  Note that DiagnonalRectangleTopology always has 4 qubits with degree 1, where as those two devices only have two qubits with degree 1.

I am okay with fixing this via name change if we can find a good name for it (I don't have a good suggestion though).

Reply:
it's true that SYC54 and bristlecone have cruft. They would be like `(4x5.5)` and `(5.5x5.5)` in this framework. The fact that they don't have complete unit cells on their boundary doesn't change the way the "bulk" grows. 

![image](https://user-images.githubusercontent.com/4967059/129422582-3766d550-c4f6-405e-ac75-206c4af5bb6a.png)